### PR TITLE
JDK-8326332: Unclosed inline tags cause misalignment in summary tables

### PR DESCRIPTION
**Pull request by hns**
Please review a simple fix to make sure inline tags are always closed in summary tables, even when the tags are closed after the first sentence in the original doc comment. 

I decided to colocate the functionality to track open inline tags with the existing `ignoreNonInlineTag` method that filters out non-inline elements because there is a lot of common functionality, such as retrieving the name and kind of an HTML tag. I considered giving the method a name that describes the additional functionality, but couldn't find one that wasn't comically long. Suggestions are welcome of course.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8326332](https://bugs.openjdk.org/browse/JDK-8326332): Unclosed inline tags cause misalignment in summary tables (**Bug** - P3)

### Reviewers
 * [Guoxiong Li](https://openjdk.org/census#gli) (@lgxbslgx - Committer) ⚠️ Review applies to [9353d1ea](https://git.openjdk.org/jdk/pull/18154/files/9353d1ea12423f6e76b90672f721e79a58e4b04f)
 * [Jonathan Gibbons](https://openjdk.org/census#jjg) (@jonathan-gibbons - **Reviewer**) ⚠️ Review applies to [9353d1ea](https://git.openjdk.org/jdk/pull/18154/files/9353d1ea12423f6e76b90672f721e79a58e4b04f)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/18154/head:pull/18154` \
`$ git checkout pull/18154`

Update a local copy of the PR: \
`$ git checkout pull/18154` \
`$ git pull https://git.openjdk.org/jdk.git pull/18154/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 18154`

View PR using the GUI difftool: \
`$ git pr show -t 18154`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/18154.diff">https://git.openjdk.org/jdk/pull/18154.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/18154#issuecomment-1983636525)

### CODE REVIEW

File: test/langtools/jdk/javadoc/doclet/testBreakIterator/TestBreakIterator.java [hunk 1]
```diff
@@ -76,5 +76,10 @@ public void test() {
                 """
                     <div class="block">A constant indicating that the keyLocation is indeterminate
                      or not relevant.</div>""");
+
+        checkOutput("pkg/BreakIteratorTest.html", true,
+                """
+                    <div class="block">Inline tags <i><a href="../index-all.html">extending
+                     beyond the first sentence.</a></i></div>""");
```
Comment (lgxbslgx) [inline test/langtools/jdk/javadoc/doclet/testBreakIterator/TestBreakIterator.java:83]: I have a question:
Why doesn't it include ` Tags are closed here`? Is it intentional or a bug?

Comment (hns) [reply test/langtools/jdk/javadoc/doclet/testBreakIterator/TestBreakIterator.java:83]: This piece of HTML is from the method summary table which only displays the first sentence of the doc comment (which is the root of the problem fixed by this PR).

Comment (lgxbslgx) [reply test/langtools/jdk/javadoc/doclet/testBreakIterator/TestBreakIterator.java:83]: Got it. Thanks for your answer.

**Review by lgxbslgx — APPROVED** [review]
LGTM.

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/HtmlDocletWriter.java [hunk 2]
```diff
@@ -1171,21 +1172,28 @@ private void addCommentTags(Element element, List<? extends DocTree> tags, boole
...
             HtmlTag htmlTag = HtmlTag.get(name);
-            if (htmlTag != null &&
-                    htmlTag.blockType != jdk.javadoc.internal.doclint.HtmlTag.BlockType.INLINE) {
-                return true;
+            if (htmlTag != null) {
+                if (htmlTag.blockType != HtmlTag.BlockType.INLINE) {
+                    return true;
+                }
+                // Keep track of open inline tags that need to be closed, see 8326332
+                if (kind == START_ELEMENT && htmlTag.endKind == HtmlTag.EndKind.REQUIRED) {
```
Comment (jonathan-gibbons) [inline src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/HtmlDocletWriter.java:1191]: Clever use of `endKind` -- I would not have thought of that.

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/HtmlDocletWriter.java [hunk 3]
```diff
@@ -1257,6 +1265,7 @@ public ContentBuilder add(CharSequence text) {
         CommentHelper ch = utils.getCommentHelper(element);
         configuration.tagletManager.checkTags(element, trees);
         commentRemoved = false;
+        ArrayList<Name> openTags = new ArrayList<>();
```
Comment (jonathan-gibbons) [inline src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/HtmlDocletWriter.java:1268]: Suggest either using `var` or just `List<Name>`.

**Review by jonathan-gibbons — APPROVED** [review]
Nice.  

I like that you managed this without needing to build anything resembling an HTML conformance checker or even to extend existing support for HTML tags.

File: src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/HtmlDocletWriter.java [hunk 4]
```diff
@@ -1171,21 +1172,28 @@ private void addCommentTags(Element element, List<? extends DocTree> tags, boole
...
-            name = eetree.getName();
+        Kind kind = dtree.getKind();
+        if (kind == Kind.START_ELEMENT) {
+            name = ((StartElementTree)dtree).getName();
+        } else if (kind == Kind.END_ELEMENT) {
+            name = ((EndElementTree)dtree).getName();
         }
 
         if (name != null) {
             HtmlTag htmlTag = HtmlTag.get(name);
```
Comment (jonathan-gibbons) [inline src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/HtmlDocletWriter.java:1185]: minor P5 suggestion to consider in future:
maybe add convenience forms for `HtmlTag.of` that accept a `DocTree`.

```
HtmlTree htmlTag = HtmlTree.of(tree);
```

```
HtmlTree of(DocTree dtree) {
    return switch (tree.getKind()) {
        case START_ELEMENT -> of(((StartElementTree)dtree).getName());
        case END_ELEMENT -> of(((StartElementTree)dtree).getName());
        default -> throw new IllegalArgumentException(tree.getKind().toString());
    }
}
```

**Pull request closed without merging**